### PR TITLE
Add primary weighs fp8 support for mxfp8

### PR DESCRIPTION
**Pull request by kunlunl**
# Description

This PR mainly adds the partial cast feature for mxfp8 primary weights. In FSDP, since each forward and backward pass requires gathering params, it's better to only gather the fp8 weights (row-wise for forward, column-wise for backward)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Add partial cast for mxfp8 primary weights

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by shjwudp** [comment]
@kunlunl Some CI attempts failed; could you please take a look?

**Comment by kunlunl** [comment]
> @kunlunl Some CI attempts failed; could you please take a look?

@shjwudp  the CI failed in building, [Errno 28] No space left on device

File: transformer_engine/common/include/transformer_engine/recipe.h [hunk 1]
```diff
@@ -122,6 +122,15 @@ void nvte_fp8_block_scaling_partial_cast(const NVTETensor inp, NVTETensor out,
...
                                          const NVTEDType out_dtype, cudaStream_t stream);
 
+void nvte_mxfp8_scaling_compute_partial_amax(const NVTETensor input, NVTETensor amax_rowwise,
+                                             NVTETensor amax_colwise, int rows, int cols,
+                                             size_t start_offset, cudaStream_t stream);
+
+void nvte_mxfp8_scaling_partial_cast(const NVTETensor input, NVTETensor output_rowwise,
+                                     NVTETensor output_colwise, const NVTETensor scale_inv_rowwise,
+                                     const NVTETensor scale_inv_colwise, int rows, int cols,
+                                     size_t start_offset, cudaStream_t stream);
```
Comment (timmoon10) [inline transformer_engine/common/include/transformer_engine/recipe.h:293]: We should document these functions. Documenting the other partial cast functions would also be nice.

File: transformer_engine/common/multi_tensor/compute_scale.cu [hunk 2]
```diff
@@ -55,6 +56,32 @@ struct ComputeScaleAndScaleInvFunctor {
   }
 };
 
+struct ComputeScaleInvE8M0Functor {
+  __device__ __forceinline__ void operator()(int chunk_size, volatile int *noop_gmem,
+                                             TensorListMetadata<2> &tl) {
+    // I'd like this kernel to propagate infs/nans.
+    // if(*noop_gmem == 1)
+    //   return;
```
Comment (timmoon10) [inline transformer_engine/common/multi_tensor/compute_scale.cu]: If we're not using the noop flag, then we shouldn't include it in `nvte_multi_tensor_compute_scale_inv_e8m0_cuda`.

File: transformer_engine/common/recipe/mxfp8_scaling.cu [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+template <typename IType>
+__global__ void __launch_bounds__(kThreadsPerBlock)
+    mxfp8_scaling_compute_partial_amax_kernel(const IType *input, IType *amax_rowwise,
+                                              IType *amax_colwise, int amax_rowwise_stride,
+                                              int amax_colwise_stride, int rows, int cols,
+                                              size_t start_offset, size_t len) {
+  __shared__ float smem_amax_rowwise[kRowsPerTile][kColsPerTile / 32];
+
+  size_t end_offset = start_offset + len;
+  const IType *input_minus_offset = input - start_offset;
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/common/recipe/mxfp8_scaling.cu:35]: @greptile-apps What's the fuss? It makes sense that accessing `input_minus_offset[idx]` with `idx<start_offset` is invalid, but that's precisely why the accesses are within an if-statement with `idx>=start_offset`. If I am reading correctly there's no correctness issue. Are you just complaining about style and risk of bugs?

File: transformer_engine/common/recipe/mxfp8_scaling.cu [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    }
+
+    if (lane_idx == 0) {
+      smem_amax_rowwise[i][warp_idx] = row_amax;
+    }
+
+    r++;
+  }
+
+  amax_colwise[blockIdx.y * amax_colwise_stride + c] = static_cast<IType>(col_amax);
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/common/recipe/mxfp8_scaling.cu:66]: @kunlunl Don't we need this to avoid writing out-of-bounds, or is this handled by the padding requirements?

File: transformer_engine/common/include/transformer_engine/recipe.h (hunk 1 above)
Comment (kunlunl) [reply transformer_engine/common/include/transformer_engine/recipe.h:293]: Added descriptions and ASCII-art diagrams.

File: transformer_engine/common/multi_tensor/compute_scale.cu (hunk 2 above)
Comment (kunlunl) [reply transformer_engine/common/multi_tensor/compute_scale.cu]: In the current implementation of the `MultiTensorApplier` https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/optimizers/multi_tensor_apply.py#L21, it is hard-coded to always pass a noop flag in to the multi tensor functor, so we need it to accept that parameter.

File: transformer_engine/common/recipe/mxfp8_scaling.cu (hunk 4 above)
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/common/recipe/mxfp8_scaling.cu:66]: Yes, it's ensured by padding requirements. Amax in mxfp8 is always padded to 128, and each CTA processes 128 columns, so there will be no out-of-bounds access. The host function `mxfp8_scaling_compute_partial_amax` also checks the amax shape is padded to correct value.

File: transformer_engine/common/recipe/mxfp8_scaling.cu (hunk 3 above)
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/common/recipe/mxfp8_scaling.cu:35]: If I don’t subtract it here, every subsequent access would have to use `input[idx - start_offset]`, so I do it once up front instead. 

The root cause is that `idx` indexes into the original logical full tensor, while input is a slice, so `input - start_offset` gives me the base address corresponding to the original logical tensor, which then lets me index directly with `idx`.

File: transformer_engine/common/recipe/mxfp8_scaling.cu [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    smem_scales_rowwise[r_][c_] = ptx::exp2f_rcp(scale_inv_rowwise[idx]);
+  }
+
+  // Load scales_colwise
+  {
+    int c_ = threadIdx.x;
+    int r = blockIdx.y * kRowsPerTile / 32;
+    int c = blockIdx.x * kColsPerTile + c_;
+    size_t idx = r * scale_inv_colwise_stride + c;
+    smem_scales_colwise[c_] = ptx::exp2f_rcp(scale_inv_colwise[idx]);
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/common/recipe/mxfp8_scaling.cu:103]: Frequent boundary checks can affect performance. I used proper CTA size (and padding size) to ensure there will be no out-of-bounds issues.

File: transformer_engine/common/recipe/mxfp8_scaling.cu [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+  __shared__ float smem_scales_colwise[kColsPerTile];
+
+  // Load scales_rowwise
+  {
+    int r_ = threadIdx.x / (kColsPerTile / 32);  // rows in shared memory
+    int c_ = threadIdx.x % (kColsPerTile / 32);  // cols in shared memory
+    int r = blockIdx.y * kRowsPerTile + r_;
+    int c = blockIdx.x * kColsPerTile / 32 + c_;
+    size_t idx = r * scale_inv_rowwise_stride + c;
+    smem_scales_rowwise[r_][c_] = ptx::exp2f_rcp(scale_inv_rowwise[idx]);
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/common/recipe/mxfp8_scaling.cu:94]: Frequent boundary checks can affect performance. I used proper CTA size (and padding size) to ensure there will be no out-of-bounds issues.

File: transformer_engine/common/recipe/mxfp8_scaling.cu [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+  amax_colwise[blockIdx.y * amax_colwise_stride + c] = static_cast<IType>(col_amax);
+
+  __syncthreads();
+
+  int r_ = threadIdx.x / (kColsPerTile / 32);  // rows in shared memory
+  int c_ = threadIdx.x % (kColsPerTile / 32);  // cols in shared memory
+  r = blockIdx.y * kRowsPerTile + r_;
+  c = blockIdx.x * kColsPerTile / 32 + c_;
+  amax_rowwise[r * amax_rowwise_stride + c] = static_cast<IType>(smem_amax_rowwise[r_][c_]);
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/common/recipe/mxfp8_scaling.cu:74]: Frequent boundary checks can affect performance. I used proper CTA size (and padding size) to ensure there will be no out-of-bounds issues.

[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/common/recipe/mxfp8_scaling.cu:74]: Frequent boundary checks can affect performance. I used proper CTA size (and padding size) to ensure there will be no out-of-bounds issues.

File: transformer_engine/common/recipe/mxfp8_scaling.cu [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+void mxfp8_scaling_compute_partial_amax(const Tensor input, Tensor amax_rowwise,
+                                        Tensor amax_colwise, int rows, int cols,
+                                        size_t start_offset, cudaStream_t stream) {
+  NVTE_CHECK(rows % 32 == 0, "rows must be divisible by 32");
+  NVTE_CHECK(cols % 32 == 0, "cols must be divisible by 32");
+
+  NVTE_CHECK(input.data.shape.size() == 1, "input must be a 1D tensor");
+  NVTE_CHECK(start_offset + input.data.shape[0] <= rows * cols, "Invalid start_offset");
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/common/recipe/mxfp8_scaling.cu]: Nice, modified.

File: transformer_engine/common/recipe/mxfp8_scaling.cu [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+void mxfp8_scaling_partial_cast(const Tensor input, Tensor output_rowwise, Tensor output_colwise,
+                                const Tensor scale_inv_rowwise, const Tensor scale_inv_colwise,
+                                int rows, int cols, size_t start_offset, cudaStream_t stream) {
+  NVTE_CHECK(rows % 32 == 0, "rows must be divisible by 32");
+  NVTE_CHECK(cols % 32 == 0, "cols must be divisible by 32");
+
+  NVTE_CHECK(input.data.shape.size() == 1, "input must be a 1D tensor");
+  NVTE_CHECK(start_offset + input.data.shape[0] <= rows * cols, "Invalid start_offset");
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/common/recipe/mxfp8_scaling.cu]: Done.

File: transformer_engine/pytorch/csrc/extensions/fp8_partial_cast.cpp [hunk 10]
```diff
@@ -48,4 +48,42 @@ void fp8_block_scaling_partial_cast(const at::Tensor &inp, at::Tensor out, const
       start_offset, block_len, static_cast<NVTEDType>(out_dtype), at::cuda::getCurrentCUDAStream());
 }
 
+void mxfp8_scaling_compute_partial_amax(const at::Tensor &input, at::Tensor amax_rowwise,
+                                        at::Tensor amax_colwise, int rows, int cols,
+                                        size_t start_offset) {
+  TORCH_CHECK(input.is_contiguous(), "input must be contiguous");
+  TORCH_CHECK(amax_rowwise.is_contiguous(), "amax_rowwise must be contiguous");
+  TORCH_CHECK(amax_colwise.is_contiguous(), "amax_colwise must be contiguous");
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/pytorch/csrc/extensions/fp8_partial_cast.cpp:56]: The check is inside `nvte_mxfp8_scaling_compute_partial_amax` and `nvte_mxfp8_scaling_partial_cast`

File: transformer_engine/pytorch/csrc/extensions/fp8_partial_cast.cpp [hunk 11]
```diff
@@ -48,4 +48,42 @@ void fp8_block_scaling_partial_cast(const at::Tensor &inp, at::Tensor out, const
...
+
+void mxfp8_scaling_partial_cast(const at::Tensor &input, at::Tensor output_rowwise,
+                                at::Tensor output_colwise, const at::Tensor &scale_inv_rowwise,
+                                const at::Tensor &scale_inv_colwise, int rows, int cols,
+                                size_t start_offset) {
+  TORCH_CHECK(input.is_contiguous(), "input must be contiguous");
+  TORCH_CHECK(output_rowwise.is_contiguous(), "output_rowwise must be contiguous");
+  TORCH_CHECK(output_colwise.is_contiguous(), "output_colwise must be contiguous");
+  TORCH_CHECK(scale_inv_rowwise.is_contiguous(), "scale_inv_rowwise must be contiguous");
+  TORCH_CHECK(scale_inv_colwise.is_contiguous(), "scale_inv_colwise must be contiguous");
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/pytorch/csrc/extensions/fp8_partial_cast.cpp:75]: The check is inside nvte_mxfp8_scaling_compute_partial_amax and nvte_mxfp8_scaling_partial_cast

File: transformer_engine/common/multi_tensor/compute_scale.cu [hunk 12]
```diff
@@ -65,6 +92,18 @@ void multi_tensor_compute_scale_and_scale_inv_cuda(int chunk_size, Tensor noop_f
...
+void multi_tensor_compute_scale_inv_e8m0_cuda(int chunk_size, Tensor noop_flag,
+                                              std::vector<std::vector<Tensor *>> tensor_lists,
+                                              cudaStream_t stream) {
+  NVTE_CHECK(tensor_lists[0][0]->data.dtype == DType::kBFloat16, "amax should be bf16");
+  auto scale_inv_dtype = tensor_lists[1][0]->data.dtype;
+  NVTE_CHECK(scale_inv_dtype == DType::kByte || scale_inv_dtype == DType::kFloat8E8M0,
+             "scale_inv should be e8m0/uint8");
+  multi_tensor_apply<2>(BLOCK_SIZE, chunk_size, noop_flag, tensor_lists,
+                        ComputeScaleInvE8M0Functor(), stream);
+  NVTE_CHECK_CUDA(cudaGetLastError());
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/common/multi_tensor/compute_scale.cu]: `MultiTensorApplier` handles the shape.

File: tests/pytorch/test_partial_cast.py
Comment (timmoon10) [inline tests/pytorch/test_partial_cast.py:1]: Do we need to create a new test script, or could we include it in `test_multi_tensor.py`? If it makes more sense to keep this file, then we need to make sure it's included in the CI script at https://github.com/NVIDIA/TransformerEngine/blob/main/qa/L0_pytorch_unittest/test.sh.

File: tests/pytorch/test_partial_cast.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+
+
+def test_mxfp8_scaling_partial_cast():
+    run_one_case(3, 32, 64, 31)
+    run_one_case(64 * 64 - 2, 64, 64, 1)
+    run_one_case(16384 * 6144, 16384, 6144, 0)
+    run_one_case(32768, 256, 128, 0)
+    run_one_case(131072, 768, 256, 0)
+    run_one_case(65536, 768, 256, 131072)
+    run_one_case(98304, 128, 768, 0)
```
Comment (timmoon10) [inline tests/pytorch/test_partial_cast.py:137]: Please write this as a Pytest script and skip when MXFP8 is not supported. Also, do we need all of these test cases? If it takes more than ~5 seconds, I would get rid of some.

File: transformer_engine/common/multi_tensor/compute_scale.cu (hunk 2 above)
Comment (timmoon10) [reply transformer_engine/common/multi_tensor/compute_scale.cu]: I see that many other multi-tensor kernels include the noop flag and don't use it. They're all deceiving and should be changed. For the time being, I've just modified this PR so the new function doesn't continue this antipattern.

File: tests/pytorch/test_partial_cast.py (hunk 13 above)
Comment (kunlunl) [reply tests/pytorch/test_partial_cast.py:137]: > Please write this as a Pytest script and skip when MXFP8 is not supported.

Done.

> Also, do we need all of these test cases? If it takes more than ~5 seconds, I would get rid of some.

It checks different corner cases of the partial amax and partial cast kernels. It only takes about 200ms to complete, so I think it won't put too much pressure on the CI system.

File: tests/pytorch/test_partial_cast.py
Comment (kunlunl) [reply tests/pytorch/test_partial_cast.py:1]: > Do we need to create a new test script, or could we include it in test_multi_tensor.py?

I think it's best to keep it if it's ok to you. Because it's not testing a multi-tensor functor, but rather two partial kernels (partial amax and partial cast). Another reason is that these two partial kernels are currently implemented in a very naive way. If we later find that they have become performance bottlenecks and need to be optimized, this test can be used to conveniently verify whether the optimized kernel is the same as the previous kernel.

> If it makes more sense to keep this file, then we need to make sure it's included in the CI script at https://github.com/NVIDIA/TransformerEngine/blob/main/qa/L0_pytorch_unittest/test.sh.

Done.
